File: app/src/main/java/com/floatingmenu/app/MainActivity.java
package com.floatingmenu.app;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity {
    private static final int SYSTEM_ALERT_WINDOW_PERMISSION = 2084;
    private static final int STORAGE_PERMISSION_CODE = 100;
    private static final int MANAGE_EXTERNAL_STORAGE_PERMISSION_CODE = 101;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.btn_start_service).setOnClickListener(v -> {
            checkPermissionsAndStart();
        });

        findViewById(R.id.btn_stop_service).setOnClickListener(v -> {
            stopService(new Intent(MainActivity.this, FloatingMenuService.class));
        });
    }

    private void checkPermissionsAndStart() {
        // 1. Check Overlay Permission
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
            askOverlayPermission();
            return;
        }

        // 2. Check Storage Permissions
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            if (!Environment.isExternalStorageManager()) {
                askManageExternalStoragePermission();
                return;
            }
        } else {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this,
                        new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE},
                        STORAGE_PERMISSION_CODE);
                return;
            }
        }

        // If both permissions are granted, start the service
        startService(new Intent(MainActivity.this, FloatingMenuService.class));
    }

    private void askOverlayPermission() {
        Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                Uri.parse("package:" + getPackageName()));
        startActivityForResult(intent, SYSTEM_ALERT_WINDOW_PERMISSION);
    }

    private void askManageExternalStoragePermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            try {
                Intent intent = new Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
                intent.addCategory("android.intent.category.DEFAULT");
                intent.setData(Uri.parse(String.format("package:%s", getApplicationContext().getPackageName())));
                startActivityForResult(intent, MANAGE_EXTERNAL_STORAGE_PERMISSION_CODE);
            } catch (Exception e) {
                Intent intent = new Intent();
                intent.setAction(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION);
                startActivityForResult(intent, MANAGE_EXTERNAL_STORAGE_PERMISSION_CODE);
            }
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == SYSTEM_ALERT_WINDOW_PERMISSION) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && Settings.canDrawOverlays(this)) {
                checkPermissionsAndStart(); // Proceed to check storage next
            } else {
                Toast.makeText(this, "Overlay Permission denied!", Toast.LENGTH_SHORT).show();
            }
        } else if (requestCode == MANAGE_EXTERNAL_STORAGE_PERMISSION_CODE) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && Environment.isExternalStorageManager()) {
                checkPermissionsAndStart(); // All done, start service
            } else {
                Toast.makeText(this, "Storage Permission denied!", Toast.LENGTH_SHORT).show();
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == STORAGE_PERMISSION_CODE) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                checkPermissionsAndStart();
            } else {
                Toast.makeText(this, "Storage Permission denied!", Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: app/src/main/java/com/floatingmenu/app/MainActivity.kt
package com.floatingmenu.app

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.core.content.ContextCompat

class MainActivity : ComponentActivity() {

    private val permissionLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
        checkAndStartService()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            MaterialTheme(colorScheme = darkColorScheme()) {
                MainScreen(
                    hasPermissions = checkPermissions(),
                    onRequestPermissions = { requestPermissions() },
                    onStartService = { startFloatingService() },
                    onStopService = { stopFloatingService() },
                    onTestShizuku = { testShizukuAccess() }
                )
            }
        }
    }

    private fun checkPermissions(): Boolean {
        val hasOverlay = Build.VERSION.SDK_INT < Build.VERSION_CODES.M || Settings.canDrawOverlays(this)
        val hasStorage = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            Environment.isExternalStorageManager()
        } else {
            ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
        }
        return hasOverlay && hasStorage
    }

    private fun requestPermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
            val intent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:$packageName"))
            permissionLauncher.launch(intent)
            return
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !Environment.isExternalStorageManager()) {
            try {
                val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
                intent.data = Uri.parse("package:$packageName")
                permissionLauncher.launch(intent)
            } catch (e: Exception) {
                val intent = Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION)
                permissionLauncher.launch(intent)
            }
            return
        }
    }

    private fun checkAndStartService() {
        if (checkPermissions()) {
            setContent {
                MaterialTheme(colorScheme = darkColorScheme()) {
                    MainScreen(
                        hasPermissions = true,
                        onRequestPermissions = { requestPermissions() },
                        onStartService = { startFloatingService() },
                        onStopService = { stopFloatingService() },
                        onTestShizuku = { testShizukuAccess() }
                    )
                }
            }
        } else {
            Toast.makeText(this, "Permissions not fully granted", Toast.LENGTH_SHORT).show()
        }
    }

    private fun startFloatingService() {
        startService(Intent(this, ComposeFloatingMenuService::class.java))
    }

    private fun stopFloatingService() {
        stopService(Intent(this, ComposeFloatingMenuService::class.java))
    }

    private val SHIZUKU_CODE = 102
    private fun testShizukuAccess() {
        if (!rikka.shizuku.Shizuku.pingBinder()) {
            Toast.makeText(this, "Shizuku is not running!", Toast.LENGTH_SHORT).show()
            return
        }
        if (rikka.shizuku.Shizuku.checkSelfPermission() != PackageManager.PERMISSION_GRANTED) {
            rikka.shizuku.Shizuku.requestPermission(SHIZUKU_CODE)
            return
        }
        try {
            val process = rikka.shizuku.Shizuku.newProcess(arrayOf("sh", "-c", "ls -l /sdcard/Android/data/com.pubg.imobile"), null, null)
            process.waitFor()
            Toast.makeText(this, "Shizuku works!", Toast.LENGTH_LONG).show()
        } catch (e: Exception) {
            Toast.makeText(this, "Shizuku failed: ${e.message}", Toast.LENGTH_LONG).show()
        }
    }
}

@Composable
fun MainScreen(
    hasPermissions: Boolean,
    onRequestPermissions: () -> Unit,
    onStartService: () -> Unit,
    onStopService: () -> Unit,
    onTestShizuku: () -> Unit
) {
    val cs = MaterialTheme.colorScheme
    Scaffold(containerColor = cs.surface) { padding ->
        Column(
            modifier = Modifier.fillMaxSize().padding(padding).padding(24.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center
        ) {
            Icon(Icons.Filled.Gamepad, contentDescription = null, modifier = Modifier.size(64.dp), tint = cs.primary)
            Spacer(modifier = Modifier.height(16.dp))
            Text("MOD V1", style = MaterialTheme.typography.headlineLarge, fontWeight = FontWeight.Bold, color = cs.onSurface)
            Text("Game Overlay Menu", style = MaterialTheme.typography.bodyLarge, color = cs.onSurfaceVariant)
            Spacer(modifier = Modifier.height(40.dp))

            if (!hasPermissions) {
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = cs.errorContainer)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text("Permissions Required", style = MaterialTheme.typography.titleMedium, color = cs.onErrorContainer)
                        Spacer(modifier = Modifier.height(8.dp))
                        Text("Overlay and Storage permissions are needed.", style = MaterialTheme.typography.bodyMedium, color = cs.onErrorContainer)
                        Spacer(modifier = Modifier.height(12.dp))
                        FilledTonalButton(onClick = onRequestPermissions, modifier = Modifier.fillMaxWidth()) {
                            Icon(Icons.Filled.Security, contentDescription = null, modifier = Modifier.size(18.dp))
                            Spacer(modifier = Modifier.width(8.dp))
                            Text("Grant Permissions")
                        }
                    }
                }
            } else {
                FilledTonalButton(onClick = onStartService, modifier = Modifier.fillMaxWidth().height(56.dp)) {
                    Icon(Icons.Filled.PlayArrow, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Start Overlay", style = MaterialTheme.typography.titleMedium)
                }
                Spacer(modifier = Modifier.height(12.dp))
                OutlinedButton(onClick = onStopService, modifier = Modifier.fillMaxWidth().height(56.dp)) {
                    Icon(Icons.Filled.Stop, contentDescription = null)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Stop Overlay", style = MaterialTheme.typography.titleMedium)
                }
                Spacer(modifier = Modifier.height(24.dp))
                TextButton(onClick = onTestShizuku) {
                    Icon(Icons.Filled.BugReport, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(modifier = Modifier.width(8.dp))
                    Text("Test Shizuku")
                }
            }
        }
    }
}
